### PR TITLE
changed mindist mapping policy specifier

### PR DESCRIPTION
**Pull request by elenash**
Сhanged mindist mapping policy specifier from map-bt dist:device,modifiers to --map-by dist:modifiers -mca rmaps_dist_device device

### CODE REVIEW

**Comment by mellanox-github** [comment]
* Coverity found 912 errors for all: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/all_376/output/errors/index.html

* Coverity found 5 errors for oshmem: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/oshmem_376/output/errors/index.html

* Coverity found 2 errors for fca: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/fca_376/output/errors/index.html <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-master-pr/376/
Test PASSed.

**Comment by rhc54** [comment]
May one ask why? In general, I'd prefer not to expand the command line yet again

**Comment by miked-mellanox** [comment]
currently, the same command line specifies device and map in one sentence
it is better two have two separate knobs to select device and mapping policy

**Comment by miked-mellanox** [comment]
+1

**Comment by rhc54** [comment]
Personally, I'd rather this not come into the code base. We continually here complaints about the large and constantly growing number of MCA params in this system. This new one doesn't do anything new at all - it just duplicates existing syntax. The argument that it is easier in a script doesn't make sense to me, as scripts have no issue with putting $vars in lists.

So a :-1:  from me barring some more convincing argument.

**Comment by miked-mellanox** [comment]
from usability perspective - having one parameter to specify many different things is too complicated.

-mca btl_tcp_if_include eth0,span

is not a user-friendly. Split into two (and keep the current syntax as well) is more user-friendly
The fix is not removing old behave, it just adds additional way to specify every component separately. can we agree on that?

**Comment by rhc54** [comment]
I'm confused - why would you put span on a BTL include param??? It's a modifier to the mapping directive.

**Comment by miked-mellanox** [comment]
it is just an example of mixing unrelated configurables in one parameter.
bad copy paste , but the point still stands.

**Comment by rhc54** [comment]
I'm still not fully understanding your point. The current parameter specifies all mapping-related directives in one place. We went to that because people were unhappy with having to specify so many separate things. How is this not a step backward?

I'd really prefer to have only _one_ way to specify things, not several ways of specifying the same thing.

**Comment by jsquyres** [comment]
I've been following this, and I confess that I'm a bit confused as well.

Here's my questions:

1. We likely already have too many MCA params.  I am usually not hesitant to add more MCA params (especially when a user asks for a specific tunable knob to turn), but I don't see the new functionality that this param adds.  What is new here?
1. I'm also usually not adverse to long names :-), but it looks like `--map-by dist:mlx4_0,span` is a whole lot fewer characters to type than `-map-by dist:mlx4_0 -mca rmaps_base_map_modifiers span`.  I'm therefore not understanding the "user-friendly" argument for this change.  Can you explain further?
1. Even though we do have (too) many MCA params, I think few (if any?) offer multiple *different* ways to get to the same functionality (aliases are the obvious exception -- but that's the *same* functionality available via multiple names).  If there are *different* ways to get to the same functionality, then you have precedence issues to consider: if multiple ways are specified, which one wins?  **NOTE:** This confusion is somewhat strengthened by the problem we're having with MCA param aliases right now -- the (slow) race condition that makes even multiple paths to the *same* functionality problematic.

**Comment by miked-mellanox** [comment]
the only reason for this PR is to simplify current form. At many customers and our internal scripts we see common usage pattern for running MPI:

* function/logic to select communication device
* function/logic to select mapper

These are two different unrelated flows, using mix of MCA or SHELL params

``` shell
-mca btl_openib_if_include $hca -x MXM_IB_DEVICE $hca ....
```

and different flow for selecting mapping method:

``` shell
--map-by $map
```

There is not correlation between two flows. 
The existing form adds this correlation and makes scripting and visual readability complicated.

Also, in mtt, we have a well-known param "alloc=..." which is device independent and should stay this way.

Mixing device and mapping selection in one sentence is not user friendly.

Now that we have 10k MCA params, we can add another two w/o hitting the overflow. 
EndUser uses 5-10 params at max, all others are internal for special tuning and it is always good to have many knobs for rainy day :)

**Comment by rhc54** [comment]
But there never will be a correlation between the two flows. Even in your proposed change, there is no correlation. You still specify the mapping policy separate from the selection of a particular interface.

The current system doesn't force any correlation. You specify your interfaces using the appropriate param. You then specify a mapping policy. If you want it to be "dist", then you have to tell us "distance from what?". If you then modify the policy, well you include that modification in the param.

This isn't a case of adding just one or two params. We'd have to add a bunch: n_per_node, n_per_socket, n_per_numa, etc. Like I said, I refuse to have multiple ways to do the same thing. So if you are going to move one modifier to a separate param, then you have to move _all_ of them.

I just don't think that makes any sense.

**Comment by hppritcha** [comment]
Mike have there been specific customer requests for this change to the way procs are mapped to the resource being specified?  

I understand your reasoning for suggesting a separation in the way the mapping resource is specified, from the way the procs are mapped to them, but given that currently there are 93 (with one deprecated) mca parameters for the rmaps framework, it might be better to consider
something more comprehensive than just adding to the list.  Also, as Jeff comments, the proposed new mca parameter has quite a long name.

**Comment by rhc54** [comment]
Where did you get this 93 number??? Running ompi_info --all thru "grep maps | wc -l" yields 51, which includes a bunch of priority, verbosity, and other non-germane things that are specific to seldom-used components..

There are exactly two MCA params today that are used by the RMAPS framework: map-by and rank-by

**Comment by miked-mellanox** [comment]
look at the mtt configuration example below. see alloc= part and suggest a nice way to exercise mindist alg w/o entering ugly zone of carrying device name (selected in different part of mtt) as part of specifying allocation policy?

Same for customer scripts.

We keep the old format as well, having two more mca params will not break any flow. Having two new params will let mtt to exercise mindist gracefully and will make many customer scripts happy and logical.


The param name can be shorter. 


``` shell
[MPI Details: OMPI]
# Check &test_alloc() for byslot or bynode
exec = mpirun @alloc@ -np &test_np() @mca@ &test_executable() &test_argv()
parameters = &MPI::OMPI::find_mpirun_params(&test_command_line(), \
                                            &test_executable())
network = &MPI::OMPI::find_network(&test_command_line(), &test_executable())

alloc = &if(&eq(&test_alloc(), "node"), "--bynode", "--byslot")

```

**Comment by rhc54** [comment]
Sounds to me like this is an MTT issue you are having, not an OMPI issue. Perhaps we should get together with Jeff and discuss the MTT issue separately?

**Comment by miked-mellanox** [comment]
mtt suffers from the same problem. it uses same rationale as a customer scripts:

select communication device flow is unrelated to the process mapping and happening in different parts of ini/scripts.

**Comment by rhc54** [comment]
Well, I'll reiterate - let's talk to Jeff about the MTT issue and see where that leads us. I still don't understand what is so hard about substituting a script variable into the MCA param (on the cmd line or in a setenv script cmd). I do it all the time.

**Comment by rhc54** [comment]
I've got a few minutes between meetings, so let me try again to explain the problem I see with this proposal. If we set this up as a separate MCA param, and leave the existing params alone, then we have _two_ ways of specifying mapping modifiers: in the map-by param, and in the "modifier" param.

The problem now is: how do I rationalize conflicts and precedence between those two params? The MCA param system cannot help me because they are two different params, so I now have to create an entirely separate logic path for dealing with it. Okay, let's try.

What do I do if the modifier param is set (e.g., to "span"), but the map-by param given in the cmd line has no modifier in it? Do I assume that the user didn't want this particular mpirun to use the modifier, and thus left it out? Or do I assume that the lack of a modifier means that the modifier param is to be used? Will the user be expecting that behavior, say if the modifier param is set out of his immediate view in the default MCA param file? How do they then indicate that they _don't_ want the modifier for this specific run?

We would have to define every one of these conflict use-cases and write logic for dealing with it. This is a non-trivial problem, and your PR doesn't deal with it.

This is why we don't have MCA params that contain overlapping functions - resolving the precedence rules becomes really hard.

HTH
Ralph

**Comment by hppritcha** [comment]
The 93 came from the output of orte_info --mca rmaps all.  The point was just to show there are indeed already a lot of parameters already for that framework.  So hence a reluctance to add more too arbitrarily.

**Comment by jsquyres** [comment]
I guess I still don't understand the use case for this PR, either.  @miked-mellanox you've cited MTT (and user scripts) -- but I think I'm failing to grok the issue: it seems like you need to specify the $device in multiple places on the command line.  Is that the problem?

**Comment by miked-mellanox** [comment]
@rhc54 - thanks for detailed info, it seems you are right on this. missed that angle. 

@jsquyres - yes, the problem is that $device needs to appear multiple times in unrelated mca params.

**Comment by jsquyres** [comment]
@miked-mellanox Ah, ok.  Is there a reason you can't assign the value to a variable and use that variable multiple times on the command line?

**Comment by miked-mellanox** [comment]
looking for user-friendly, elegant and visually simple solution which will not lead to re-write many existing scripts and mtt.ini files to adopt. (select device and mapper are unrelated flows, where some of them are loaded from outside like modules or functions.sh files or mtt.ini)

Also, --map-by dist:nic,modifier is too overloaded way to express mapping.

Do you think it can have simple form, like;

--map-by-dev $hca --map-by node

anyway, Ralph is right in his argument, will close it.

**Comment by rhc54** [comment]
What if you split the device out into a separate MCA param for that component? In other words, something like this:

--map-by dist:modifier -mca rmaps_dist_device $hca

**Comment by miked-mellanox** [comment]
this could work too! thanks! we will go for it.

**Comment by miked-mellanox** [comment]
@elenash - could you please rework this PR to support: 
--map-by dist:modifier -mca rmaps_dist_device $hca

**Comment by elenash** [comment]
Done. But this change should be documented somehow, I guess. Users used to set device after the colon, now modifiers are being set there,  it can lead to misunderstanding what is going wrong.

**Comment by elenash** [comment]
PR for jenkins to cope with this change
https://github.com/mellanox-hpc/jenkins_scripts/pull/4

**Comment by miked-mellanox** [comment]
can we support both? detect if well-known modifier goes after dist or treat it as HCA if unknown?

**Comment by mellanox-github** [comment]
* Coverity found 917 errors for all: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/all_402/output/errors/index.html

* Coverity found 5 errors for oshmem: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/oshmem_402/output/errors/index.html

* Coverity found 2 errors for fca: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/fca_402/output/errors/index.html <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-master-pr/402/
Test PASSed.

**Comment by elenash** [comment]
I don't think it's a good idea because when a user just put a wrong specifier name, it will be treated as hca device and user will never know what was the reason of the error.

**Comment by miked-mellanox** [comment]
yep, sounds right

**Comment by miked-mellanox** [comment]
updated jenkins with test for mindist new flags
bot:retest

**Comment by elenash** [comment]
bot:retest

**Comment by mellanox-github** [comment]
* Coverity found 923 errors for all: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/all_403/output/errors/index.html

* Coverity found 5 errors for oshmem: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/oshmem_403/output/errors/index.html

* Coverity found 2 errors for fca: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/fca_403/output/errors/index.html <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-master-pr/403/
Test PASSed.

**Comment by mellanox-github** [comment]
* Coverity found 923 errors for all: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/all_405/output/errors/index.html

* Coverity found 5 errors for oshmem: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/oshmem_405/output/errors/index.html

* Coverity found 2 errors for fca: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/fca_405/output/errors/index.html <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-master-pr/405/
Test PASSed.

**Comment by miked-mellanox** [comment]
:+1:

**Comment by elenash** [comment]
@rhc54 @jsquyres How do you think this change should be documented? There is not enough information about how to use --map-by dist in man pages. As I remember, there was just a post about it in Cisco blog.
